Source/player.cpp

```diff
@@ -4802,164 +4802,59 @@ void __fastcall InitDungMsgs(int pnum)
-				{
-					v8 = plr[v1]._pClass;
-					sfxdelay = 40;
-					if ( v8 )
-					{
-						if ( v8 == 1 )
-						{
-							sfxdnum = PS_ROGUE98;
-						}
-						else if ( v8 == 2 )
-						{
-							sfxdnum = PS_MAGE98;
-						}
-					}
-					else
-					{
-						sfxdnum = PS_WARR98;
-					}
-					v4 = v7 | 4;
-					goto LABEL_14;
-				}
-			}
-			break;
-		case 13u:
-			v1 = v0;
-			if ( !plr[v1]._pLvlVisited[13] && gbMaxPlayers == 1 )
-			{
-				v9 = plr[v1].pDungMsgs;
-				if ( !(v9 & 8) )
-				{
-					v10 = plr[v1]._pClass;
-					sfxdelay = 40;
-					if ( v10 )
-					{
-						if ( v10 == 1 )
-						{
-							sfxdnum = PS_ROGUE99;
-						}
-						else if ( v10 == 2 )
-						{
-							sfxdnum = PS_MAGE99;
-						}
-					}
-					else
-					{
-						sfxdnum = PS_WARR99;
-					}
-					v4 = v9 | 8;
-					goto LABEL_14;
-				}
-			}
-			break;
-		case 16u:
-			v1 = v0;
-			if ( !plr[v1]._pLvlVisited[15] && gbMaxPlayers == 1 )
-			{
-				v11 = plr[v1].pDungMsgs;
-				if ( !(v11 & 0x10) )
-				{
-					v12 = plr[v1]._pClass;
-					sfxdelay = 40;
-					if ( !v12 || v12 == 1 || v12 == 2 )
-						sfxdnum = PS_DIABLVLINT;
-					v4 = v11 | 0x10;
-					goto LABEL_14;
-				}
-			}
-			break;
+
+	if ( currlevel == 1 && !plr[myplr]._pLvlVisited[1] && gbMaxPlayers == 1 && !(plr[myplr].pDungMsgs & DMSG_CATHEDRAL) ) {
+		sfxdelay = 40;
+		if ( plr[myplr]._pClass == PC_WARRIOR ) {
+			sfxdnum = PS_WARR97;
+		} else if ( plr[myplr]._pClass == PC_ROGUE ) {
+			sfxdnum = PS_ROGUE97;
+		} else if ( plr[myplr]._pClass == PC_SORCERER ) {
+			sfxdnum = PS_MAGE97;
+		}
+		plr[myplr].pDungMsgs = plr[myplr].pDungMsgs | DMSG_CATHEDRAL;
+	} else if ( currlevel == 5 && !plr[myplr]._pLvlVisited[5] && gbMaxPlayers == 1 && !(plr[myplr].pDungMsgs & DMSG_CATACOMBS) ) {
+		sfxdelay = 40;
+		if ( plr[myplr]._pClass == PC_WARRIOR ) {
+			sfxdnum = PS_WARR96B;
+		} else if ( plr[myplr]._pClass == PC_ROGUE ) {
+			sfxdnum = PS_ROGUE96;
+		} else if ( plr[myplr]._pClass == PC_SORCERER ) {
+			sfxdnum = PS_MAGE96;
+		}
+		plr[myplr].pDungMsgs |= DMSG_CATACOMBS;
+	} else if ( currlevel == 9 && !plr[myplr]._pLvlVisited[9] && gbMaxPlayers == 1 && !(plr[myplr].pDungMsgs & DMSG_CAVES) ) {
+		sfxdelay = 40;
+		if ( plr[myplr]._pClass == PC_WARRIOR ) {
+			sfxdnum = PS_WARR98;
+		} else if ( plr[myplr]._pClass == PC_ROGUE ) {
+			sfxdnum = PS_ROGUE98;
+		} else if ( plr[myplr]._pClass == PC_SORCERER ) {
+			sfxdnum = PS_MAGE98;
+		}
+		plr[myplr].pDungMsgs |= DMSG_CAVES;
+	} else if ( currlevel == 13 && !plr[myplr]._pLvlVisited[13] && gbMaxPlayers == 1 && !(plr[myplr].pDungMsgs & DMSG_HELL) ) {
+		sfxdelay = 40;
+		if ( plr[myplr]._pClass == PC_WARRIOR ) {
+			sfxdnum = PS_WARR99;
+		} else if ( plr[myplr]._pClass == PC_ROGUE ) {
+			sfxdnum = PS_ROGUE99;
+		} else if ( plr[myplr]._pClass == PC_SORCERER ) {
+			sfxdnum = PS_MAGE99;
+		}
+		plr[myplr].pDungMsgs |= DMSG_HELL;
+	} else if ( currlevel == 16 && !plr[myplr]._pLvlVisited[15] && gbMaxPlayers == 1 && !(plr[myplr].pDungMsgs & DMSG_DIABLO) ) {

```

Comment:
This function also shows how haphazard some of the code is, with every single statment checking if we are in single player instead of just checking once in the beginning. The check for both _pLvlVisited and pDungMsgs is probably to avoid having the message play when doing a second play threw. Personally I was always sad that it didn't play in MP and I don't see a reason not to have it on a second play threw. It's not like the player shows much awareness any way and quest dialogue is still played when encountered again. Silly game designers :P

Reply:
Pretty sure they get reset for each new game(playthrough), it's just that they only play the first time you enter each stage. It'd be a bit odd if your character announced "The sanctity of this place has been fouled" every single time you entered the church.

Reply:
In that case it's odd that they would be checking both if the player has been to that level and a special flag that only apples to this function :/